Code/Engine/Core/Scripting/Implementation/ScriptComponent.cpp

```diff
@@ -0,0 +1,302 @@
+{
+  SUPER::OnActivated();
+
+  CallScriptFunction(ezComponent_ScriptBaseClassFunctions::OnActivated);
+}
+
+void ezScriptComponent::OnDeactivated()
+{
+  SUPER::OnDeactivated();
+
+  CallScriptFunction(ezComponent_ScriptBaseClassFunctions::OnDeactivated);
+}
+
+void ezScriptComponent::OnSimulationStarted()
+{
+  SUPER::OnSimulationStarted();
+
+  CallScriptFunction(ezComponent_ScriptBaseClassFunctions::OnSimulationStarted);
+}
+
+void ezScriptComponent::BroadcastEventMsg(ezEventMessage& msg)
+{
+  const ezRTTI* pType = msg.GetDynamicRTTI();
+  for (auto& sender : m_EventSenders)
+  {
+    if (sender.m_pMsgType == pType)
+    {
+      sender.m_Sender.SendEventMessage(msg, this, GetOwner());
+      return;
+    }
+  }
+
+  auto& sender = m_EventSenders.ExpandAndGetRef();
+  sender.m_pMsgType = pType;
+  sender.m_Sender.SendEventMessage(msg, this, GetOwner());
+}
+
+void ezScriptComponent::SetScriptClass(const ezScriptClassResourceHandle& hScript)
+{
+  if (m_hScriptClass == hScript)
+    return;
+
+  ClearInstance(true);
+
+  m_hScriptClass = hScript;
+
+  if (IsActiveAndInitialized() && m_hScriptClass.IsValid())
+  {
+    InstantiateScript(true);
+  }
+}
+
+void ezScriptComponent::SetScriptClassFile(const char* szFile)
+{
+  ezScriptClassResourceHandle hScript;
+
+  if (!ezStringUtils::IsNullOrEmpty(szFile))
+  {
+    hScript = ezResourceManager::LoadResource<ezScriptClassResource>(szFile);
+  }
+
+  SetScriptClass(hScript);
+}
+
+const char* ezScriptComponent::GetScriptClassFile() const
+{
+  return m_hScriptClass.IsValid() ? m_hScriptClass.GetResourceID().GetData() : "";
+}
+
+void ezScriptComponent::SetUpdateInterval(ezTime interval)
+{
+  m_UpdateInterval = interval;
+
+  if (IsActiveAndInitialized())
+  {
+    UpdateScheduling();
+  }
+}
+
+ezTime ezScriptComponent::GetUpdateInterval() const
+{
+  return m_UpdateInterval;
+}
+
+const ezRangeView<const char*, ezUInt32> ezScriptComponent::GetParameters() const
+{
+  return ezRangeView<const char*, ezUInt32>([]() -> ezUInt32 { return 0; },
+    [this]() -> ezUInt32 { return m_Parameters.GetCount(); },
+    [](ezUInt32& it) { ++it; },
+    [this](const ezUInt32& it) -> const char* { return m_Parameters.GetKey(it).GetString().GetData(); });
+}
+
+void ezScriptComponent::SetParameter(const char* szKey, const ezVariant& value)
+{
+  ezHashedString hs;
+  hs.Assign(szKey);
+
+  auto it = m_Parameters.Find(hs);
+  if (it != ezInvalidIndex && m_Parameters.GetValue(it) == value)
+    return;
+
+  m_Parameters[hs] = value;
+}
+
+void ezScriptComponent::RemoveParameter(const char* szKey)
+{
+  if (m_Parameters.RemoveAndCopy(ezTempHashedString(szKey)))
+  {
+  }
+}
+
+bool ezScriptComponent::GetParameter(const char* szKey, ezVariant& out_value) const
+{
+  ezUInt32 it = m_Parameters.Find(szKey);
+
+  if (it == ezInvalidIndex)
+    return false;
+
+  out_value = m_Parameters.GetValue(it);
+  return true;
+}
+
+void ezScriptComponent::InstantiateScript(bool bActivate)
+{
+  ClearInstance(true);
```

Comment:
wouldn't the bool parameter depend on whether OnActivated was called before ?

Reply:
Indeed.